### PR TITLE
Fix serialization for timestamps with precision higher than 3 for MySQL

### PR DESCRIPTION
**Pull request by hashhar**
Fixes https://github.com/trinodb/trino/issues/6852.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1241,7 +1241,7 @@
             <dependency>
                 <groupId>mysql</groupId>
                 <artifactId>mysql-connector-java</artifactId>
-                <version>5.1.48</version>
+                <version>8.0.23</version>
```
Comment (findepi) [inline pom.xml:1244]: We don't have to upgrade the driver to fix the timestamp precision issue along, so this is actually 2 separate changes and go by 2 separate PRs. The reason for this being MySQL/MariaDB compatibility issues (unless we already lost MariaDB compatibility). And MariaDB compatibility may be tied to AWS Aurora compatibility as well.

**Comment by hashhar** [comment]
I'll need to disable some test-cases because we currently use `SqlTimestamp` while writing which means that timestamps lying within a _gap_ do not round-trip.

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java [hunk 2]
```diff
@@ -310,6 +321,18 @@ protected String getTableSchemaName(ResultSet resultSet)
...
+
+            case Types.TIMESTAMP:
+                // TODO support higher precisions
+                TimestampType timestampType = TIMESTAMP_MILLIS;
+                return Optional.of(ColumnMapping.longMapping(
+                        timestampType,
+                        (resultSet, columnIndex) -> {
+                            Timestamp timestamp = resultSet.getTimestamp(columnIndex);
+                            Instant instant = timestamp.toLocalDateTime().atZone(UTC).toInstant();
+                            return toPrestoTimestamp(timestampType, LocalDateTime.ofEpochSecond(instant.getEpochSecond(), Math.toIntExact(round(instant.getNano(), 9 - timestampType.getPrecision())), UTC));
```
Comment (findepi) [inline plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:333]: The _default_ read function is not correct -- it should either round OR fail, but should not return incorrect data silently as it does today.

File: plugin/trino-mysql/src/test/java/io/trino/plugin/mysql/TestMySqlTypeMapping.java [hunk 3]
```diff
@@ -485,11 +488,122 @@ public void testDatetime()
...
+//                .addRoundTrip("TIMESTAMP '1969-12-31 23:59:59.999999949999'", "TIMESTAMP '1969-12-31 23:59:59.9999999'")
+//                .addRoundTrip("TIMESTAMP '1969-12-31 23:59:59.99999994'", "TIMESTAMP '1969-12-31 23:59:59.9999999'");
+
+        Session session = Session.builder(getSession())
+                .setTimeZoneKey(TimeZoneKey.getTimeZoneKey(sessionZone.getId()))
+                .build();
+
+        tests.execute(getQueryRunner(), session, trinoCreateAsSelect(session, "test_timestamp"));
+        tests.execute(getQueryRunner(), session, trinoCreateAsSelect(getSession(), "test_timestamp"));
+        tests.execute(getQueryRunner(), session, trinoCreateAndInsert(session, "test_timestamp"));
```
Comment (findepi) [inline plugin/trino-mysql/src/test/java/io/trino/plugin/mysql/TestMySqlTypeMapping.java:592]: Add  atest when data is written with MySQL and read with Trino -- the existing tests should not exercise the read path's rounding, as the data should be stored pre-rounded on the Trino side.

**Comment by findepi** [comment]
> I'll need to disable some test-cases because we currently use SqlTimestamp while writing which means that timestamps lying within a gap do not round-trip.

Please do not disable them completely. just use a different input and expected output literals, with a clear TODO comment.
This will ensure we update the test when we fix underlying problem, while the comment ensures no-one will get confused about that.

cc @nineinchnick

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java (hunk 2 above)
Comment (hashhar) [reply plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:333]: I don't think it returns incorrect data - a VerifyException is thrown if the value and the precision don't match.

The issue here was that the value returned from MySQL has precision != 3 and hence the `verify` fails. It's our fault as callers of the code to not provide the correct `TimestampType` (we pass `TIMESTAMP_MILLIS` always).

Should be addressed as part of https://github.com/trinodb/trino/issues/6910.

**Comment by hashhar** [comment]
> > I'll need to disable some test-cases because we currently use SqlTimestamp while writing which means that timestamps lying within a gap do not round-trip.
> 
> Please do not disable them completely. just use a different input and expected output literals, with a clear TODO comment.

Sorry for the misleading comment, there are no tests today - I meant disabling the ones I was trying to add. Instead I'll add the full set of tests (and fixes) as part of https://github.com/trinodb/trino/issues/6910.

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java (hunk 2 above)
Comment (findepi) [reply plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:333]: > I don't think it returns incorrect data - a VerifyException is thrown if the value and the precision don't match.

Right. That's a new addition (#6654), and so #6852 reported a different stacktrace than one we would get today.

Still, we could add an opt-in mode where the value is rounded (can be eg `boolean round`)
This should be used by all connectors which use TIMESTAMP_MILLIS (not only MySQL, but also the legacy default mappings (#6607) and maybe some others.

**Comment by findepi** [comment]
> Sorry for the misleading comment, there are no tests today - I meant disabling the ones I was trying to add. 

That's exactly what i understood.

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java [hunk 4]
```diff
@@ -310,6 +318,20 @@ protected String getTableSchemaName(ResultSet resultSet)
...
             case Types.DATE:
                 return Optional.of(dateColumnMapping());
+
+            case Types.TIMESTAMP:
+                // support higher precisions (TODO https://github.com/trinodb/trino/issues/6910)
+                TimestampType timestampType = TIMESTAMP_MILLIS;
+                return Optional.of(ColumnMapping.longMapping(
+                        timestampType,
+                        (resultSet, columnIndex) -> {
+                            Timestamp timestamp = resultSet.getTimestamp(columnIndex);
```
Comment (dain) [inline plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:328]: BTW, if the driver is JDBC 4.2 compatible, then you should be able use use:
```
Instant instant = resultSet.getObject(columnIndex, Instant.class);
```

**Review by jirassimok — COMMENTED** [review]
Looks mostly good. My most important comments are the last two.

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java (hunk 4 above)
Comment (hashhar) [reply plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:328]: We'd need to upgrade the MySQL driver version - which I tried but I need to add some more tests to make sure we don't break anything (notably MariaDB) in the process.

Comment (jirassimok) [reply plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:328]: Are we really using that old a version?

Anyway, it looks like MySQL's `getObject(i, LocalDateTime.class)` just calls `getTimestamp(i).toLocalDateTime()`, so that should work just as well.

Comment (hashhar) [reply plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:328]: The version we are on does't have handling for any `java.time` parameters to `getObject`.

Also, a notable bug https://bugs.mysql.com/bug.php?id=93444 (so I'll need to bump to a version containing this fix once I do bump the version).

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java (hunk 2 above)
Comment (hashhar) [reply plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java]: Note that we need to propagate the `boolean` through `legacyToPrestoType` -> `timestampColumnMapping` -> `toPrestoTimestamp`.

I think it's more worthwhile to add the `boolean` (or a method with more meaningful name) to `toPrestoTimestamp` itself and add explicit column mapping for timestamps to connectors which are using `legacyToPrestoType`.

I would like to do this in a separate PR though.

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java [hunk 5]
```diff
@@ -310,6 +318,20 @@ protected String getTableSchemaName(ResultSet resultSet)
...
+                            Instant instant = timestamp.toLocalDateTime().atZone(UTC).toInstant();
+                            return toPrestoTimestamp(timestampType, LocalDateTime.ofEpochSecond(instant.getEpochSecond(), toIntExact(round(instant.getNano(), 9 - timestampType.getPrecision())), UTC));
```
[the remark this answers is not in the document]
Comment (hashhar) [reply plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:330]: Good catch but `Instant.ofEpochSecond` does that for me.

[the remark this answers is not in the document]
Comment (hashhar) [reply plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:330]: Good catch. Thanks. Looks like some of the rounding tests were causing values to be rounded on MySQL end itself so tests weren't able to catch it. I've adjusted the tests to account for it.

**Comment by hashhar** [comment]
AC @findepi @jirassimok 

Thanks @jirassimok for catching the rounding/overflow bug.

**Comment by hashhar** [comment]
CI hit https://github.com/trinodb/trino/issues/5758.

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java [hunk 6]
```diff
@@ -310,6 +316,23 @@ protected String getTableSchemaName(ResultSet resultSet)
...
+                TimestampType timestampType = TIMESTAMP_MILLIS;
+                return Optional.of(ColumnMapping.longMapping(
+                        timestampType,
+                        (resultSet, columnIndex) -> {
+                            LocalDateTime timestamp = resultSet.getTimestamp(columnIndex).toLocalDateTime();
+                            int roundedNanos = toIntExact(round(timestamp.getNano(), 9 - timestampType.getPrecision()));
+                            LocalDateTime roundedTimestamp = timestamp
+                                    .withNano(roundedNanos % toIntExact(NANOSECONDS_PER_SECOND))
+                                    .plusSeconds(roundedNanos / NANOSECONDS_PER_SECOND);
+                            return toPrestoTimestamp(timestampType, roundedTimestamp);
```
Comment (findepi) [inline plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:331]: The need for rounding applies to MemSQL, Druid, Redshift connectors as well -- i.e. caller of https://github.com/trinodb/trino/blob/922bfa1a32cc2e39d24d79f9801099938614dd82/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java#L634 who do not handle timestamps on their own.


That flow used to round until be2b97a8480f657f311e673fa38867c8faf4e2df, but it went unnoticed until we added more verifications in other places.

MySQL is not a special case here, and we have other connectors suffering -- those who're not smart enough to handle their mapping by themsleves. We should fix the mapping returned from `io.trino.plugin.jdbc.StandardColumnMappings#jdbcTypeToPrestoType` by rounding to millis. This should be an opt-in, ie it must not be applied to `timestampColumnMappingUsingSqlTimestamp` or `timestampColumnMapping` when someone calls them directly.

**Review by findepi — CHANGES_REQUESTED** [review]

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java (hunk 6 above)
Comment (findepi) [reply plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:331]: continued @ https://github.com/trinodb/trino/pull/6893#discussion_r576641252

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java (hunk 2 above)
Comment (findepi) [reply plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java]: > Note that we need to propagate the boolean through legacyToPrestoType -> timestampColumnMapping -> toPrestoTimestamp.

`legacyToPrestoType` should not take this boolean. This is the place where we make decision to map to TS_MILLIS. 
But yes, you'd need to pass a boolean to `timestampColumnMappingUsingSqlTimestamp` and on.

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java (hunk 6 above)
Comment (hashhar) [reply plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:331]: Makes sense. I'll change the mapping returned from `legacyToPrestoType` while leaving the type-level mappings (e.g. `timestampColumnMapping` as it is.

Comment (hashhar) [reply plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:331]: With rounding enabled all `verifyPredicate` calls fail for values where we round during reads - most likely because to predicate generated in `verifyPredicate` is being pushed down.

I'll need to disable pushdown for all connectors using the `legacyToPrestoType`.

**Comment by hashhar** [comment]
@findepi AC but notice that pushdown is now disabled for anyone using the legacy timestamp mappings.

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java [hunk 7]
```diff
@@ -319,20 +314,7 @@ protected String getTableSchemaName(ResultSet resultSet)
 
             case Types.TIMESTAMP:
                 // TODO support higher precisions (https://github.com/trinodb/trino/issues/6910)
-                TimestampType timestampType = TIMESTAMP_MILLIS;
-                return Optional.of(ColumnMapping.longMapping(
-                        timestampType,
-                        (resultSet, columnIndex) -> {
-                            LocalDateTime timestamp = resultSet.getTimestamp(columnIndex).toLocalDateTime();
-                            int roundedNanos = toIntExact(round(timestamp.getNano(), 9 - timestampType.getPrecision()));
-                            LocalDateTime roundedTimestamp = timestamp
```
Comment (findepi) [inline plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java]: Did you mean to squash this with the first commit?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java [hunk 8]
```diff
@@ -434,21 +435,30 @@ public static LongWriteFunction timeWriteFunction(int precision)
     /**
      * @deprecated This method uses {@link java.sql.Timestamp} and the class cannot represent date-time value when JVM zone had
      * forward offset change (a 'gap'). This includes regular DST changes (e.g. Europe/Warsaw) and one-time policy changes
-     * (Asia/Kathmandu's shift by 15 minutes on January 1, 1986, 00:00:00). If driver only supports {@link LocalDateTime}, use
-     * {@link #timestampColumnMapping} instead.
+     * (Asia/Kathmandu's shift by 15 minutes on January 1, 1986, 00:00:00). This mapping also disables pushdown by default
+     * to ensure correctness because rounding happens within Trino and won't apply to remote system.
+     * If driver only supports {@link LocalDateTime}, use {@link #timestampColumnMapping} instead.
```
Comment (jirassimok) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java:440]: If you're editing this javadoc, also remove the "only" on this line.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java [hunk 9]
```diff
@@ -434,21 +435,30 @@ public static LongWriteFunction timeWriteFunction(int precision)
...
      */
     @Deprecated
-    public static ColumnMapping timestampColumnMappingUsingSqlTimestamp(TimestampType timestampType)
+    public static ColumnMapping timestampColumnMappingUsingSqlTimestampWithRounding(TimestampType timestampType)
```
Comment (jirassimok) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java:443]: I'm not a fan of this name. I think we can leave it unchanged. If we really want a name with all that information in it, I think `roundingSqlTimestampMapping` or `roundingSqlTimestampColumnMapping` would be better.

File: plugin/trino-mysql/src/test/java/io/trino/plugin/mysql/TestMySqlTypeMapping.java [hunk 10]
```diff
@@ -478,18 +481,256 @@ public void testDate()
...
+                // negative epoch (MySQL's timestamp type doesn't support values <= epoch)
+                //.addRoundTrip("TIMESTAMP '1969-12-31 23:59:59.99999995'", "TIMESTAMP '1970-01-01 00:00:00.000'")
+                //.addRoundTrip("TIMESTAMP '1969-12-31 23:59:59.999999949999'", "TIMESTAMP '1969-12-31 23:59:59.999'")
+                //.addRoundTrip("TIMESTAMP '1969-12-31 23:59:59.99999994'", "TIMESTAMP '1969-12-31 23:59:59.999'")
+
+                .execute(getQueryRunner(), session, mysqlCreateAndInsert("tpch.test_timestamp"));
+    }
+
+    @DataProvider
+    public Object[][] sessionZonesDataProvider()
```
Comment (jirassimok) [inline plugin/trino-mysql/src/test/java/io/trino/plugin/mysql/TestMySqlTypeMapping.java:723]: You can make this return `ZoneId[]` to make the method body look a little nicer.

One of IntelliJ's TestNG inspections ("invalid data provider return type") doesn't like it very much, but IntelliJ is wrong here.

File: plugin/trino-mysql/src/test/java/io/trino/plugin/mysql/TestMySqlTypeMapping.java [hunk 11]
```diff
@@ -478,18 +481,253 @@ public void testDate()
...
+
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.1'", "TIMESTAMP '2020-09-27 12:34:56.100'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.9'", "TIMESTAMP '2020-09-27 12:34:56.900'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.123'", "TIMESTAMP '2020-09-27 12:34:56.123'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.123000'", "TIMESTAMP '2020-09-27 12:34:56.123'")
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.999'", "TIMESTAMP '2020-09-27 12:34:56.999'")
+                // max supported precision
+                .addRoundTrip("TIMESTAMP '2020-09-27 12:34:56.123456'", "TIMESTAMP '2020-09-27 12:34:56.123'")
+
+                // round down
```
[the remark this answers is not in the document]
Comment (jirassimok) [reply plugin/trino-mysql/src/test/java/io/trino/plugin/mysql/TestMySqlTypeMapping.java:565]: This also works for `addRoundTrip(String, String, Type, String)`, even if the type makes it look a little less nice.

```java
.addRoundTrip(
        "datetime(6)",
        "TIMESTAMP '1970-01-01 00:00:01.1234569'",
        TIMESTAMP_MILLIS,
        "TIMESTAMP '1970-01-01 01:00:01.123'")
```

File: plugin/trino-mysql/src/test/java/io/trino/plugin/mysql/TestMySqlTypeMapping.java (hunk 10 above)
Comment (findepi) [reply plugin/trino-mysql/src/test/java/io/trino/plugin/mysql/TestMySqlTypeMapping.java:723]: I agree IntelliJ is wrong and its inspection needs an update, but i don't want to have to eye-ball & ignore warnings manually.
So unless we decide the inspection is not helpful and we recommend disabing it for everyone (see DEVELOPING.md), we should avoid triggering warnings.

**Comment by hashhar** [comment]
AC as fixup.

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java [hunk 12]
```diff
@@ -310,6 +311,10 @@ protected String getTableSchemaName(ResultSet resultSet)
 
             case Types.DATE:
                 return Optional.of(dateColumnMapping());
+
+            case Types.TIMESTAMP:
+                // TODO support higher precisions (https://github.com/trinodb/trino/issues/6910)
+                return Optional.of(timestampColumnMappingUsingSqlTimestampWithRounding(TIMESTAMP_MILLIS));
```
Comment (findepi) [inline plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java]: OK to remove these line for now, as the intention is to fix the default mapping.

```suggestion
                break;
```

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java [hunk 13]
```diff
@@ -434,21 +435,30 @@ public static LongWriteFunction timeWriteFunction(int precision)
...
     {
         // TODO support higher precision
         checkArgument(timestampType.getPrecision() <= TimestampType.MAX_SHORT_PRECISION, "Precision is out of range: %s", timestampType.getPrecision());
         return ColumnMapping.longMapping(
                 timestampType,
                 (resultSet, columnIndex) -> {
-                    Timestamp timestamp = resultSet.getTimestamp(columnIndex);
-                    return toPrestoTimestamp(timestampType, timestamp.toLocalDateTime());
+                    LocalDateTime timestamp = resultSet.getTimestamp(columnIndex).toLocalDateTime();
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java:450]: timestamp -> localDateTime
roundedTimestamp -> rounded

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java [hunk 14]
```diff
@@ -434,21 +435,30 @@ public static LongWriteFunction timeWriteFunction(int precision)
...
+                    int roundedNanos = toIntExact(round(timestamp.getNano(), 9 - timestampType.getPrecision()));
+                    LocalDateTime roundedTimestamp = timestamp
+                            .withNano(roundedNanos % toIntExact(NANOSECONDS_PER_SECOND))
+                            .plusSeconds(roundedNanos / NANOSECONDS_PER_SECOND);
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java:454]:
```
                            .withNano(0)
                            .plusNanos(roundedNanos);
```

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java [hunk 15]
```diff
@@ -445,16 +447,18 @@ public static ColumnMapping timestampColumnMappingUsingSqlTimestamp(TimestampTyp
...
-
-    @Deprecated
-    public static ColumnMapping timestampColumnMapping()
-    {
-        return timestampColumnMapping(TIMESTAMP_MILLIS);
+                timestampWriteFunctionUsingSqlTimestamp(timestampType),
+                // NOTE: pushdown is disabled because the values stored in remote system might not match the values as
+                // read by Trino due to rounding. This can lead to incorrect results if operations are pushed down to
+                // the remote system.
+                DISABLE_PUSHDOWN);
```
Comment (findepi) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/StandardColumnMappings.java:461]: I liked the rename of this method to `timestampColumnMappingUsingSqlTimestampWithRounding` to indicate the semantics change. It's safe, since there are no other usages. It doesn't need to be pretty, since it's deprecated and likely to go away at some point.

File: plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java [hunk 16]
```diff
@@ -310,6 +311,10 @@ protected String getTableSchemaName(ResultSet resultSet)
 
             case Types.DATE:
                 return Optional.of(dateColumnMapping());
+
+            case Types.TIMESTAMP:
+                // TODO support higher precisions (https://github.com/trinodb/trino/issues/6910)
+                return Optional.of(timestampColumnMappingUsingSqlTimestamp(TIMESTAMP_MILLIS));
```
Comment (findepi) [inline plugin/trino-mysql/src/main/java/io/trino/plugin/mysql/MySqlClient.java:317]:
```suggestion
                break; // currently handled by the default mappings
```

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
As a followup we should add same tests for MemSQL and Druid.

**Comment by hashhar** [comment]
AC as fixups again. Will squash the fixups after approval.

PTAL again @findepi @jirassimok

**Review by findepi — APPROVED** [review]

**Comment by hashhar** [comment]
Squashed the fixups. Should be good to merge on a green build.

**Comment by findepi** [comment]
Please rebase to force CI run

**Comment by findepi** [comment]
@hashhar see CI
